js/general_patterns.md

```diff
@@ -0,0 +1,210 @@
+# General Patterns
+
+This document provides an overview of the general architecture and programming
+patterns that will be used throughout the JavaScript API. These patterns have
+been chosen because of the flexibility they provide in integrating with native
+APIs and ease to build new interfaces from, and for their ability to create
+performant and more optimizable code.
+
+
+
+### Asynchronous Calls
+
+The asynchronous callback pattern is very similar to `setTimeout()`. With the
+exception that required arguments must always go first. After the callback, N
+arguments can be passed that will propagate to the callback.
+
+Example of pattern:
+```js
+asyncCall(<arguments>, callback[, ...args])
+```
+
+`<arguments>` are the arguments required by the function call.
+
+`callback` is the function that is executed when the asynchronous operation is
+complete.
+
+`...args` are the additional values that will propagate to the callback.
```

Comment:
What is the motivation to allow an arbitrary number of arguments to be passed that propagate to the callback?

When no arguments can propagate:

1) A user is forced to nest callbacks in order to maintain access to specific variables. This inhibits optimization and forces pyramid code.
OR
2) A user must use `Function#bind()`, clobbering the `this` of the callback and horribly inhibiting performance.
OR
3) Properties must be attached to the return instance that can be retrieved via `this.*` in the callback. Thus altering the object map which reduces performance and potentially allows object properties to be overridden.

When allowing only one argument to propagate:

4) User must create an object or an array if they wish multiple values to be passed. This is less aesthetic for the user, and I haven't seen *noticeable performance impact allowing arbitrary number of arguments.

 *Based on the fact it's unlikely they will pass more than 4 arguments.

So this hits the sweet spot in terms of usability and performance. While currently it requires a small shim to do correctly, in the future [restful parameters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters) may help with this. (Though that is conditional upon whether all function parameters are required or not. If they aren't then an array operation will have to take place regardless.)